### PR TITLE
Remove fast-json-stringify and allows customJSONSerializer to be a function

### LINKED ISSUE #83: Make fast-json-stringify optional
Thanks for the great project.

**Is your feature request related to a problem? Please describe.**

I'm experimenting with using `graphql-jit` in the browser. Currently `graphql-jit` is bundled with `fast-json-stringify`, which makes its size extremely big to be used in browser.

https://bundlephobia.com/result?p=graphql-jit

Removing `fast-json-stringify` would cut the size in half.

**Describe the solution you'd like**

Currently, we have the option `compilerOptions.customJSONSerializer`, which accepts a boolean value. However the option does not provide much use because it only switches between `JSON.stringify` and `fastJSON`, not really 'custom'.

My proposal is that we replace it with `compilerOptions.fastJSON`, which will either be `undefined` or a user-supplied `require('fast-json-stringify')`.

### PR DESCRIPTION
**Pull request by hoangvvo**
This PR allows `options.customJSONSerializer` to be a function to be called with `CompilationContext` to returns a serializer function. It will be a breaking change.

Something like `queryToJSONSchema` in [json.ts](https://github.com/zalando-incubator/graphql-jit/blob/master/src/json.ts) is not a concern of this package and should be reimplemented in userland.

```js
import { queryToJSONSchema } from '../user-land/json';

compileQuery(schema, document, "", {
  customJSONSerializer: (context) => {
    const jsonSchema = queryToJSONSchema(context);
    return fastJson(jsonSchema);
  }
});
```

### Why

- `fast-json-stringify` is way too large if `graphql-jit` is ever used in browser.
- We pin `fast-json-stringify` at v1 which misses out a lot of features from v2. User should get to choose what JSON serializer to use.

Fix #83

### CODE REVIEW

**Comment by hoangvvo** [comment]
I tried to publish it as a fork. Bundle size went down from `226.9 kb` -> `70.3 kb` (~70%)

https://bundlephobia.com/result?p=@hoangvvo/graphql-jit@0.5.0

compared to 

https://bundlephobia.com/result?p=graphql-jit@0.4.3

File: README.md [hunk 1]
```diff
@@ -110,7 +110,7 @@ Compiles the `document` AST, using an optional operationName and compiler option
     so this option should only be set to true if there are strong assurances that the values are valid.
   - `customSerializers` {Object as Map, default: {}} - Replace serializer functions for specific types. Can be used as a safer alternative
     for overly expensive serializers
-  - `customJSONSerializer` {boolean, default: false} - Whether to produce also a JSON serializer function using `fast-json-stringify`. The default stringifier function is `JSON.stringify`
+  - `customJSONSerializer` {function, default: false} - A function to be called with [`CompilationContext`](https://github.com/zalando-incubator/graphql-jit/blob/master/src/execution.ts#L87) to produce also a JSON serializer function. The default stringifier function is `JSON.stringify`
```
Comment (ruiaraujo) [inline README.md:113]:
```suggestion
  - `customJSONSerializer` {function, default: undefined} - A function to be called with [`CompilationContext`](https://github.com/zalando-incubator/graphql-jit/blob/master/src/execution.ts#L87) to produce also a JSON serializer function. The default stringifier function is `JSON.stringify`
```

**Comment by ruiaraujo** [comment]
Isn't this equivalent to removing the feature fully?

File: README.md (hunk 1 above)
Comment (hoangvvo) [reply README.md:113]: In code, the default value is `false`. Is this okay to write it as `undefined` in the doc (even though it does not matter actually)

https://github.com/zalando-incubator/graphql-jit/blob/8b678a10b9ec1e16db84b0566f4d64fe9be9b5de/src/execution.ts#L214

**Comment by hoangvvo** [comment]
Sorry for the late response. It is helpful to me, but I'm not sure if others are depending on it (in which case let's drop this PR).

By the name `customJSONSerializer`, the option should allow actually "custom" JSON serializer, not specifically the built in one.

Pretty much my reasons have already been mentioned above in this PR. This feature seems to be "out of the scope" of this package.

**Comment by ruiaraujo** [comment]
@boopathi what do you think about this? We could move to a mutiple package setup and add the JSON stringifier with the idea of later adding a protobuf one for example.

**Comment by boopathi** [comment]
yes I think that's a good idea. we can add and move a few other things as well.

**Comment by hoangvvo** [comment]
Are we still interested in proceeding in this direction? I would love to reopen another PR after discussing the good approach forward. Thanks!

**Pull request closed without merging**